File: leetcode/solved/x_of_a_kind_in_a_deck_of_cards.rb
# @param {Integer[]} deck
# @return {Boolean}
def has_groups_size_x(deck)
  sieve = Array.new(200, true)
  primes = []
  for p in (2..100)
    next unless sieve[p]
    primes.push(p)

    i = p*p
    while i <= 10000
      sieve[i] = false
      i += p
    end
  end

  freq = {}
  deck.each do |c|
    freq[c] ||= 0
    freq[c] += 1
  end

  primes.each do |p|
    ans = true
  	freq.each do |k,v|
      ans &= (v % p == 0)
    end

    return true if ans
  end

  return false
end

puts has_groups_size_x([1,1,1,2,2,2,2,2,2])

File: leetcode/solved/x_of_a_kind_in_a_deck_of_cards.cc
class Solution {
public:
    bool hasGroupsSizeX(vector<int>& deck) {
      if(deck.size() < 2)
        return false;

      vector<bool> sieve(100, true);
      vector<int> primes;
      for(int p=2; p<=100; ++p) {
        if(sieve[p] == false) continue;
        primes.push_back(p);

        for(int i=p*p; i<=100; i+=p)
          sieve[i] = false;
      }

      map<int,int> freq;
      for(vector<int>::iterator i=deck.begin(); i!=deck.end(); ++i)
        ++freq[*i];

      for(vector<int>::iterator p=primes.begin(); p!=primes.end(); ++p) {
        bool ans = true;
        for(map<int,int>::iterator i=freq.begin(); i!=freq.end(); ++i)
          ans &= (i->second % *p == 0);

        if(ans == true)
          return true;
      }

      return false;
    }
};